sdk/include/opentelemetry/sdk/trace/span_data.h

```diff
@@ -76,6 +86,11 @@ class SpanData final : public Recordable
     parent_span_id_ = parent_span_id;
   }
 
+  void SetAttribute(nostd::string_view key, common::AttributeValue &&value) noexcept override
+  {
+    attributes_[std::string(key)] = value;
```

Comment:
Just to confirm - do we intend to make this thread safe or the caller has to take care of it?

Reply:
FYI - https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/concurrency.md

Reply:
More info:
* In Python (to be specific, CPython) this is [inherently thread safe](https://docs.python.org/3/glossary.html#term-global-interpreter-lock).
* In C# this is done by [CAS (compare and swap)](https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs#L311) (note: `Activity.AddTag` provides the same semantic as `Span.SetAttribute`.

Reply:
Well, that answers it:
> **Span** - All methods of Span are safe to be called concurrently.


I'll add some locking around `attributes_` and I will submit another PR making all other `SpanData` operations atomic.

I'll also add some documentation to `Recordable`, as this is something everybody implementing their own `Recordable` has to take into consideration.

Reply:
You can safely ignore my last comment, this is handled on the [level of the `Span` in the SDK](https://github.com/open-telemetry/opentelemetry-cpp/pull/117/files#diff-ff684db3f68fbab2b9eb31f749753c1dR75).

`Recordable`s themselves need not be thread safe, `Span` ensures appropriate locking around access to the `Recordable`.

Reply:
What would happen if previous attribute already exists and is of a different type than the new attribute being set? SetAttribute("key", <string_view>) then SetAttribute("key", bool)

Reply:
The `string_view` variant would be overwritten by a `bool` variant. A similar case is covered by [this unit test](https://github.com/open-telemetry/opentelemetry-cpp/pull/117/files#diff-3ca430cf0ba066c7e55415eccabfa5adR125).